src/cms/preview-templates/lib/createPreview.jsx

```diff
@@ -4,47 +4,75 @@ import { useState, useLayoutEffect } from 'react'
 
 export default function createPreview(component) {
   return (props) => {
-    setUpPreviewPane(props)
-    moveStylesToPreviewPane(props)
+    const { window, document } = props
+
+    useLayoutEffect(() => {
+      const cleanUpId = setPreviewPaneId(document)
+      const cleanUpFontAwesome = setUpFontAwesome(document)
+      const cleanUpClickHandler = ignoreLinksAndButtons(document)
+
+      return () => {
+        cleanUpId()
+        cleanUpFontAwesome()
+        cleanUpClickHandler()
+      }
+    }, [document])
```

Comment:
Try to use custom hooks or react components as the units of composability in React code. `setPreviewPaneId`, `setUpFontAwesome` and `ignoreLinksAndButtons` are hook-like in how they're implemented, so let's just go ahead and make them hooks.

```ts
function useAppId(document) {
  useLayoutEffect(() => {
    const oldId = document.body.getAttribute("id")
    document.body.setAttribute("id", "app")

    return () => document.body.setAttribute("id", oldId)
  }, [document])
}

function useFontAwesome(document) {
  useLayoutEffect(() => {
    const fontAwesome = document.createElement('script')
    fontAwesome.setAttribute('src', 'https://kit.fontawesome.com/74c5cf378a.js')
    fontAwesome.setAttribute('crossorigin', 'anonymous')
    document.head.appendChild(fontAwesome)

    return () => document.head.removeChild(fontAwesome)
  }, [document])
}

function useIgnoreClicks(document) {
  useEffect(() => {
    const handler = event => {
      if (["A", "BUTTON"].includes(event.target.tagName)) {
        event.preventDefault()
        event.stopPropagation()
      }
    }

    document.body.addEventListener("click", handler)

    return () => document.body.removeEventListener("click", handler)
  }, [document])
}
```

Then you can compose these hooks to make your component

```ts
return (props) => {
  const { window, document } = props

  useAppId(document)
  useFontAwesome(document)
  useIgnoreClicks(document)

  // ...
}
```

React simply uses the naming convention of `useX` to determine if a function is a custom hook or not, but conforming to the convention will allow linter rules to check your logic appropriately.

Reply:
Note that I used `useEffect()` for the click handler, but left the others as `useLayoutEffect`. Generally if the effect is going to change what the browser paints, we want to use useLayoutEffect for that, but effects that don't impact the paint (such as adding event listeners) should be useEffect instead, so they're attached after the paint.

Strictly speaking, I don't think it matters if useFontAwesome() is a layout effect or not, because I doubt the runkit is going to parse the page and load the icons in time for the first paint to be impacted by it, but I suppose it's aspirational to have it be a layout effect. Like "we WANT this to happen before the paint" even though the reality of how the script works probably makes that unrealistic.

Reply:
I've made the functions back into hooks, and got rid of `useLayoutEffect` since none of these functions affect the layout in any important way.